plutus-core/docs/equalsCanonical.md

```diff
@@ -0,0 +1,299 @@
+
+They use the same structural equality function as a "canonical example of run-time type analysis".
+
+Overall, a structural equality operator does complicate metatheory, but it doesn't destroy it.
+
+It is however important to look at the whole system, which in our case comprises four languages, not just TPLC and PIR.
+
+## Security of the system as a whole
+
+If `equalsCanonical : all a. a -> a -> bool` is still admissible, do we win anything at all by introducing the `mkCanonical` indirection? Yes: we introduce a hoop that the user needs to jump through to be able to use `equalsCanonical` and that hoop acts as explicit guidance.
+
+We have a concrete example of such a thing in the code base: the [`Normalized`](https://github.com/input-output-hk/plutus/blob/99afebc7a0ce5352dd40fb86bac6255003028fd7/plutus-core/plutus-core/src/PlutusCore/Core/Type.hs#L194) newtype wrapper. Its constructor is exposed, because there's simply no sensible way not to expose it as one often has to manually construct `Normalized` types in the type checker and some safe automatically normalizing function can't be used for performance reasons. So while it doesn't provide any safety guarantees, it's still useful, because it points the developer to every place where attention is required when it comes to not breaking the internal invariant that type inference returns a `Normalized` type. It is then the responsibility of the developer to explain every usage of the `Normalized` constructor in the type checker and we do enforce that policy during code reviews.
+
+`mkCanonical` plays the same role, with the user normally not being somebody who creates a smart contract, but a compiler developer, because we're not going to expose the unsafe `mkCanonical` builtin directly for smart contract writers to use. Instead, in Plutus Tx we'll expose something like
+
+```haskell
+mkCanonical :: ToData a => a -> Canonical a
+```
+
+which, given that `ToData` instances are normally derived, is mostly safe with the exception of the user manually providing a `ToData` instance with a non-injective `toBuiltinData` (which may even make sense for a quotient type for example).
+
+As an option we could introduce a `MkCanonical` type class whose purpose is to ensure that every call to `mkCanonical` is safe, instead of relying on `toBuiltinData` being injective for all types. Using `ToData` has the advantage that with a flag one can switch from structural equality to equality-via-`Data` to test that the two agree.
+
+Whatever we choose in the end to help the user not to shoot themselves in the foot with Plutus Tx version of `mkCanonical`, it needs to come with some sort of a "soft proof" like a satisfied instance constraint. By converting this "soft proof" to a `a -> data` function we create a true ghost of a departed proof, except rather than being a ghost within the same system like in the paper, in our case it is a ghost from a higher-level one.
```

Comment:
Or even further: on the Haskell side, since the class constraint is taking the place of the `canonical` argument, we could just have
```
class Canonical a where
   equalsCanonical :: a -> a -> Bool
```
with a similar setup

Reply:
> I don't really see the advantage of using a -> data over a "soft" satisfied instance constraint.

There's no advantage to `a -> data` when the satisfied instance constraint exists. The point of `a -> data` is exactly to be reminiscent of that instance down the language tower.

> `instance TypeError => MkCanonical (a -> b)`

Yes, but you can't catch, say, type variables.

>  plus ideally some kind of deriving that gives you an instance iff all your fields have one.

It should be in the proposal, yes.

Reply:
> Yes, but you can't catch, say, type variables.

I'm not sure I see the problem there. `MkCanonical a => ...` is problematic if `a` can be a function type, but we're not going to allow instances for function types, so it can never be?